lib/tokenizer/preprocessor.js

```diff
@@ -26,15 +31,34 @@ var Preprocessor = module.exports = function () {
 
     this.pos = -1;
     this.lastGapPos = -1;
+    this.lastCharPos = -1;
+    this.droppedBufferSize = 0;
 
     this.gapStack = [];
 
     this.skipNextNewLine = false;
 
     this.lastChunkWritten = false;
     this.endOfChunkHit = false;
+    this.bufferWaterline = DEFAULT_BUFFER_WATERLINE;
 };
 
+Object.defineProperty(Preprocessor.prototype, 'sourcePos', {
+    get: function () {
+        return this.droppedBufferSize + this.pos;
+    }
+});
+
+Preprocessor.prototype.dropParsedChunk = function () {
+    if (this.pos > this.bufferWaterline) {
```

Comment:
We will need to introduce special flag that will signal that we met hibernation token which we will need to check in initial states and then reset it. So we have a lot of duplicate code or we'll need to introduce additional method call (by refactoring flag check + preprocessor.dropParsedChunk + flag reset into separate method of Tokenizer). Also, with small chunks we may have opposite effect, since we'll produce a lot of intermediate strings. Overall, I'm not sure about this approach to be honest.